drv/lpc55-update-server/src/images.rs

```diff
@@ -23,82 +31,605 @@ extern "C" {
         match (component, slot) {
-            (RotComponent::Hubris, SlotId::A) => (
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-            ),
-            (RotComponent::Hubris, SlotId::B) => (
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+            (RotComponent::Hubris, SlotId::A) => FlashRange {
+                store: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+                exec: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+            },
+            (RotComponent::Hubris, SlotId::B) => FlashRange {
+                store: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+                exec: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+            },
+            (RotComponent::Stage0, SlotId::A) => FlashRange {
+                store: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
+            },
+            (RotComponent::Stage0, SlotId::B) => FlashRange {
+                store: __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+            },
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC format.
+impl From<&[u8]> for Epoch {
+    fn from(chars: &[u8]) -> Self {
+        match chars.iter().try_fold(0u32, |epoch: u32, c| {
+            if let Some(digit) = (*c).checked_sub(b'0') {
+                if digit < 10 {
+                    epoch
+                        .checked_mul(10u32)
+                        .and_then(|product| product.checked_add(digit as u32))
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }) {
+            Some(value) => Epoch { value },
+            None => Epoch::from(0u32),
+        }
+    }
+}
```

Comment:
I know you said it makes it easier for you to reason about but I do not find this easy to review. I would really rather see us use a standard library function as a first pass and then write our own if we can't actually deal with the panics.

Reply:
I'm fine with that.